common-fate/granted
# Add AWS IAM Identity Center device code flow automation

### What changed?
Adds support for automatically confirming the user code used in the IAM Identity Center login flow. Here's a demo: https://www.loom.com/share/2ecdc053a0bd474dbccfec9d9fc084ab.

When authenticating to IAM Identity Center, Granted now stores the user code in the keychain. The new [Granted Chrome extension](https://chromewebstore.google.com/detail/granted/cjjieeldgoohbkifkogalkmfpddeafcm) communicates with Granted via the [Native messaging API](https://developer.chrome.com/docs/extensions/develop/concepts/native-messaging).

This feature is currently supported on Chromium-based browsers. We need to update our Firefox extension to support it.

### Why?
This offers a better UX when using IAM Identity Center, and is more secure as it mitigates the risk of [AWS SSO phishing](https://ramimac.me/aws-device-auth).

If the device code *doesn't* match one that Granted recognises (i.e. it was sent to you by an attacker trying to phish your AWS IAM Identity Center credentials), the Granted browser extension will disable the login button:

<img width="1840" alt="Screenshot 2024-09-19 at 19 36 10" src="https://github.com/user-attachments/assets/a2a44e9a-e2b5-411c-a825-def992f7177a">


### How did you test it?
Confirmed working with a local build of Granted, with the production build of the Chrome extension.

Test process:

0. Ensure you have configured Chrome as your default SSO browser with Granted (`granted browser set-sso`)
1. Install the [Granted browser extension](https://chromewebstore.google.com/detail/granted/cjjieeldgoohbkifkogalkmfpddeafcm)
2. Build the CLI: `sudo make cli`
3. Clear the token cache to ensure a fresh login is required: `dgranted cache clear --all`
3. Run `dassume` on a role which uses IAM Identity Center: `dassume`

Granted should confirm the user code shown in the browser automatically.

### Potential risks
May impact IAM Identity Center login flow but code change here is quite small, risk of breaking the login flow if the user's keychain is misconfigured (although in that case, Granted itself probably won't work either).

### Is patch release candidate?
No, new feature

### Link to relevant docs PRs
Docs TBC
